Headless chrome v2

deivid-rodriguez opened the pull request:
#### :tophat: What? Why?

This starts the migration to headless chrome. We can't fully migrate yet, since headless chrome still can't manipulate windows and some tests need that. So I left poltergeist around for those tests. I haven't played around with it but there's some additions to poltergeist that might no longer be necessary. But first I want to get a green build. 

#### :pushpin: Related Issues
- Hopefully this helps with #1425.

#### :clipboard: Subtasks
- [x] Add headless chrome.
- [x] Migrate what can be migrated to headless chrome.
- [ ] Remove the stuff no longer needed for the leftover `poltergeist` tests.
- [x] <strikeout>Move downloaded binaries to user's home bin folder (which is already in the path in Circle CI).</strikeout> Don't really need them in the path, selenium takes options for the paths to both binaries.
- [ ] Either go back to circle v2.0 or bring back parallelism, yarn test && artifacts to v1.0 config.
- [x] <strikeout>Spit out to the console somehow the chromium build the test run is using (or even freeze it). This is very likely to break, so we need at least an easy way to revert back to a known good revision.</strikeout> The download script already spits this, so it's fine.

### :camera: Screenshots (optional)
_None_.

#### :ghost: GIF
![headless](https://cloud.githubusercontent.com/assets/2887858/26799710/0d741c5e-4a0d-11e7-9c06-8585afa693c9.gif)

deivid-rodriguez commented:
Can't believe it actually passed!

Still needs some cleanup though:

* [ ] Bring back parallelism and all the `yarn test` stuff.
* [ ] Move downloaded binaries to user's home bin folder (which is already in the path in Circle CI).
* [ ] Bring back artifacts.

I had to downgrade to Circle v1.0 because headless chrome hits https://bugs.chromium.org/p/chromedriver/issues/detail?id=1772 in the v2.0 environment. I don't think it's a big deal for now.

mrcasals commented:
Summoning @beagleknight here.

@deivid-rodriguez can't we keep CircleCI 2.0 and add XVFB meanwhile?

beagleknight commented:
@mrcasals @deivid-rodriguez For CircleCI 2.0 I think we can use a remote selenium using the official [selenium image](https://hub.docker.com/r/selenium/standalone-chrome/). I'm not sure if it's support headless but we can give it a try.

Or we can just install xvfb 🤷‍♂️ and keep it there until it is no longer needed.

deivid-rodriguez commented:
What's the advantage of v2.0 though? I kind of dislike using different software for CI and local builds. And, well, v2.0 is still a beta, I feel we are using too many beta things at once, maybe? :) Anyways, if we can get that working I'm up for it. Also, I briefly considered using xvfb for _both_ local and CI runs, but then I read that it doesn't work on OSX (and I assumed you're using that, are you?).

deivid-rodriguez commented:
Also feel free to contribute to this branch! Just don't force-push the current head since it's the only green full build we have... xD

josepjaume commented:
@deivid-rodriguez what about the build matrix though? We only have 1x parallelism now (instead of 4x)

deivid-rodriguez commented:
It's in the TODO list! :wink:

deivid-rodriguez commented:
The lightning fast build we have now is priceless. We should _definitely_ keep that.

beagleknight commented:
@deivid-rodriguez The main advantage of CircleCi 2.0 is we can use custom containers so we can run our tests using the `decidim` docker image so our environment is deterministic and we are not tied to CircleCi environment changes.
I know it is in beta but to me it is pretty stable and I think the beta will end very soon!

But you are right about mirroring the local environment too. I think we can document the project to run the tests using a remote selenium with docker, as I suggested to do in Circle, and manual instructions to install Chrome (with headless support) and xvbf. I think both solutions work well in Mac OSX (although Docker is slower in OSX systems).

I will give it a try to Circle 2.0 tomorrow :)!

deivid-rodriguez commented:
Sounds like a good plan to me @beagleknight! And yes, I agree CircleCI v2.0 is very stable and works great.

The tests themselves are ready I think so now it 's a matter of getting the infrastructure right!

deivid-rodriguez commented:
@beagleknight Funny enough, v1.0 does not seem 100% stable either! :)

![captura](https://user-images.githubusercontent.com/2887858/26839156-958a6fde-4ab8-11e7-84cd-8e3555dff646.png)

deivid-rodriguez commented:
I rebased this and cleaned it up a bit.

I reenabled parallelism but the build is still using a single processor. I think this is because v1.0 does not allow configuring parallelism through `circle.yml` but only in the UI. Anyways, I'll now let @beagleknight give this a try in v2.0 (maybe @mrcasals can help too since he configured circle CI initially). Hopefully we can get this working with all the docker niceties!

Once we get a green parallel build, we can clean this up further by removing `poltergeist` tweaks, one at a time.

beagleknight commented:
@deivid-rodriguez I lost a battle this morning against Docker and my Mac OS X 😭 . I am not sure if I can make it work...
I also found this https://bugs.chromium.org/p/chromedriver/issues/detail?id=1772#c12
It could be nice if we get rid off xvfb...

deivid-rodriguez commented:
Hehe, yes. I linked to that issue in the first comment of the PR, since it's the reason why I had to fallback to circle v1.0.

deivid-rodriguez commented:
> It could be nice if we get rid off xvfb...

I didn't get this, we're not using that at the moment, right?

I think we have two options at this point:

* Try out "headfull" chrome in circle v2.0 with xvfb.
* Leave it as it is with headless chrome in circle v1.0.

I'd go with the latter since it's already implemented. But if we were to try the former, I'd switch our dev environments to use `xvfb` as well. Otherwise the power of a consistent environment provided by `docker` is a bit lost if we use different tools locally vs in CI.

beagleknight commented:
@deivid-rodriguez Well... I'm sure Circle 1.0 have `xvfb` installed because Chrome headless can't work without it right now 😄 ! The "problem" with Circle 2.0 is we need to setup the environment either installing the dependencies on the container running the tests or using external containers.
I wanted to use the default `selenium/standalone-chrome` but I can't get it work in my dev machine (because of Mac OS X duh...)

deivid-rodriguez commented:
> @deivid-rodriguez Well... I'm sure Circle 1.0 have xvfb installed because Chrome headless can't work without it right now :smile: 

Really? I'm not using `xvfb` and tests pass for me. I thought native headless mode was precisely replacing the need for `xvfb`... Also note that both me and circle use Linux, maybe that's related.

beagleknight commented:
@deivid-rodriguez Well... I am a bit confused then 😵  `Xvfb` is supposed to be mandatory even using headless mode if we use the `sendKeys` command.

deivid-rodriguez commented:
Well, in any case, I don't think this is a blocker any more since #1458! :)

deivid-rodriguez commented:
I'm going to close this for now. Once headless chrome plays nice with our dev environments & circle CI v2.0 without the need for xfvb, we can reopen. Third time's the charm they say! :)
